### ISSUE TITLE
java mhap thread numbers

### ISSUE THREAD

**Issue by imneuro** [marbl/canu#1144]
Hi Canu team,

I have been using canu-1.6 to assemble plant genome for a while with slurm as schedular. The command is default canu setting. Recently I notice that several mhap jobs mysteriously disappeared from compute nodes. I happen to recorded the  job command and processes before one of the job disappeared. 

I observe unusual a the process tree of PID61966. It shows over 80 threads for this job, yet on command line it only requested 12 threads. I want your advice on if this is what it suppose to be. If not, could this be a reason to cause the job disappear? Any suggestion on further trouble-shooting or decrease the threads number?

```
root     61956     1  0 13:55 ?        00:00:00 slurmstepd: [27553] 
root     61960 61956  0 13:55 ?        00:00:00 /bin/bash /tmp/slurmd/job27553/slurm_script 0 
root     61966 61960 99 13:55 ?        02:11:59 java -d64 -server -Xmx22528m -jar /opt/canu/bin/mhap-2.1.2.jar --repeat-weight 0.9 --repeat-idf-scale 10 -k 16 --num-hashes 256 --num-min-matches 3 --threshold 0.8 --filter-threshold 0.000005 --ordered-sketch-size 1000 --ordered-kmer-size 14 --min-olap-length 500 --num-threads 12 ...
```

```
$ top 
top - 14:11:54 up 49 days, 15:55,  1 user,  load average: 59.65, 51.97, 42.70 
Tasks: 724 total,   3 running, 721 sleeping,   0 stopped,   0 zombie 
%Cpu(s): 77.1 us,  3.1 sy,  0.0 ni, 19.8 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st 
KiB Mem : 14415649+total, 21733788 free, 61420892 used, 61001812 buff/cache 
KiB Swap:        0 total,        0 free,        0 used. 81826088 avail Mem 

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND 
61966 root      20   0 29.638g 0.013t  15352 S  1174  9.4 128:59.08 java 
62335 root      20   0 29.638g 0.011t  15472 S  1172  7.9  59:38.38 java 
62444 root      20   0 29.638g 0.010t  15628 S  1163  7.4  66:51.94 java 
62578 root      20   0 29.638g 0.010t  15460 S  1131  7.8  62:46.34 java 
62149 root      20   0 29.638g 0.013t  15448 S  1122  9.6  89:26.91 java 
62548 root       0 -20       0      0      0 R   2.3  0.0   0:09.08 kworker/19:1H
``` 

```$ top 

top - 14:11:54 up 49 days, 15:55,  1 user,  load average: 59.65, 51.97, 42.70 
Tasks: 724 total,   3 running, 721 sleeping,   0 stopped,   0 zombie 
%Cpu(s): 77.1 us,  3.1 sy,  0.0 ni, 19.8 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st 
KiB Mem : 14415649+total, 21733788 free, 61420892 used, 61001812 buff/cache 
KiB Swap:        0 total,        0 free,        0 used. 81826088 avail Mem 

PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND 
61966 root      20   0 29.638g 0.013t  15352 S  1174  9.4 128:59.08 java 
62335 root      20   0 29.638g 0.011t  15472 S  1172  7.9  59:38.38 java 
62444 root      20   0 29.638g 0.010t  15628 S  1163  7.4  66:51.94 java 
62578 root      20   0 29.638g 0.010t  15460 S  1131  7.8  62:46.34 java 
62149 root      20   0 29.638g 0.013t  15448 S  1122  9.6  89:26.91 java 
62548 root       0 -20       0      0      0 R   2.3  0.0   0:09.08 kworker/19:1H 
```
Here is the job scripts
```$tail -n 100 /tmp/slurmd/job27567/slurm_script 
 
if [ $jobid -eq 771 ] ; then 
  blk="000309" 
  slf="" 
  cvt="-h 20328001 0 -q 20328001" 
  qry="000771" 
fi 

if [ x$qry = x ]; then 
  echo Error: Job index out of range. 
  exit 1 
fi 

if [ -e ./results/$qry.ovb ]; then 
  echo Job previously completed successfully. 
  exit 
fi 

#  File must exist: queries.tar 

if [ -e ./queries.tar -a ! -d ./queries ] ; then 
  tar -xf ./queries.tar 
fi 

if [ ! -d ./results ]; then 
  mkdir -p ./results 
fi 

if [ ! -d ./blocks ] ; then 
  mkdir -p ./blocks 
fi 

#  File must exist: blocks/$blk.dat 
for ii in `ls ./queries/$qry` ; do 
  echo Fetch blocks/$ii 
#  File must exist: blocks/$ii 
done 

#  File must exist: BEBA693.ms16.frequentMers.ignore.gz 

echo "" 
echo Running block $blk in query $qry 
echo "" 

if [ ! -e ./results/$qry.mhap ] ; then 
  java -d64 -server -Xmx22528m \ 
    -jar  $bin/mhap-2.1.2.jar  \ 
    --repeat-weight 0.9 --repeat-idf-scale 10 -k 16 \ 
    --num-hashes 256 \ 
    --num-min-matches 3 \ 
    --threshold 0.8 \ 
    --filter-threshold 0.000005 \ 
    --ordered-sketch-size 1000 \ 
    --ordered-kmer-size 14 \ 
    --min-olap-length 500 \ 
    --num-threads 12 \ 
    -s  ./blocks/$blk.dat $slf  \ 
    -q  queries/$qry  \ 
  > ./results/$qry.mhap.WORKING \ 
  && \ 
  mv -f ./results/$qry.mhap.WORKING ./results/$qry.mhap 
fi 


if [   -e ./results/$qry.mhap -a \ 
     ! -e ./results/$qry.ovb ] ; then 
  $bin/mhapConvert \ 
    -G ../BEBA693.gkpStore \ 
    $cvt \ 
   -o ./results/$qry.mhap.ovb.WORKING \ 
    ./results/$qry.mhap \ 
  && \ 
  mv ./results/$qry.mhap.ovb.WORKING ./results/$qry.mhap.ovb 
fi 
 

if [   -e ./results/$qry.mhap -a \ 
       -e ./results/$qry.mhap.ovb ] ; then 
  rm -f ./results/$qry.mhap 
fi 
 

if [ -e ./results/$qry.mhap.ovb ] ; then 
  mv -f ./results/$qry.mhap.ovb ./results/$qry.ovb 
fi 

#  File is important: results/$qry.ovb 
#  File is important: results/$qry.counts 

  

exit 0 
```

**Comment by skoren** [comment]
Nothing in Canu would make the jobs disappear, you'd have to check your scheduler's history for the slurm jobs in question, perhaps it was killed for exceeding requested memory/threads?

The threads are probably JVM system threads, like garbage collection or similar, see issue #993 for suggested java options to disallow this. This was fixed in the code but version 1.6 is quite old so wouldn't have this fix.

**Comment by skoren** [comment]
Idle, JVM threads are limited in latest code + the top report in this issue indicates the appropriate number of threads were being used.
